### PR TITLE
Client Partition Fetch converted to Listener

### PR DESCRIPTION
**Pull request by sancar**
For backward compatibility purposes old way is not removed.

ported from https://github.com/hazelcast/hazelcast/pull/10976

compatibility tests are here https://github.com/hazelcast/hazelcast-enterprise/pull/1630

### CODE REVIEW

**Comment by sancar** [comment]
run-lab-run

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java [hunk 1]
```diff
@@ -438,12 +438,12 @@ public void start() {
         proxyManager.init(config, clientContext);
         listenerService.start();
         loadBalancer.init(getCluster(), config);
-        partitionService.start();
         statistics.start();
         clientExtension.afterStart(this);
     }
 
     public void onClusterConnect(Connection ownerConnection) throws Exception {
+        partitionService.listenPartitionTable(ownerConnection);
```
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java]: what happens if event is received between connection established and this point? is there a way to register the listener earlier?

File: hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientPartitionServiceImpl.java [hunk 2]
```diff
@@ -66,21 +70,47 @@ public ClientPartitionServiceImpl(HazelcastClientInstanceImpl client) {
...
-        // use internal execution service for all partition refresh process (do not use the user executor thread)
+    public void listenPartitionTable(Connection ownerConnection) {
+        //scheduling left in place to support server versions before 3.9.
         clientExecutionService.scheduleWithRepetition(new RefreshTask(), INITIAL_DELAY, PERIOD, TimeUnit.SECONDS);
+
+        //Servers after 3.9 supports listeners
+        ClientMessage clientMessage = ClientAddPartitionListenerCodec.encodeRequest();
+        ClientInvocation invocation = new ClientInvocation(client, clientMessage, ownerConnection);
+        invocation.setEventHandler(this);
+        invocation.invokeUrgent();
```
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientPartitionServiceImpl.java]: When talking to 3.9 and later, should we fail if the invocation result is a failure? and For earlier than 3.9, we probably do not need to check the invocation result as done here.

File: hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientPartitionServiceImpl.java [hunk 3]
```diff
@@ -127,14 +147,12 @@ private boolean getPartitions() {
...
 
-    private ClientInvocationFuture getPartitionsFrom(Connection connection) {
+    private ClientInvocationFuture invokeGetPartitions() {
         ClientMessage requestMessage = ClientGetPartitionsCodec.encodeRequest();
-        return new ClientInvocation(client, requestMessage, connection).invokeUrgent();
+        return new ClientInvocation(client, requestMessage).invokeUrgent();
     }
 
-    private boolean processPartitionResponse(ClientGetPartitionsCodec.ResponseParameters response) {
-        logger.finest("Processing partition response.");
```
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientPartitionServiceImpl.java]: do we not need this log anymore?

File: hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientPartitionServiceImpl.java [hunk 4]
```diff
@@ -226,14 +244,8 @@ public void run() {
                 return;
             }
 
-            Connection connection = getOwnerConnection();
```
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientPartitionServiceImpl.java]: Removing this check may cause periodic printing of log at line 240. Assume no cluster connection, and consider what happens.

File: hazelcast-client/src/test/java/com/hazelcast/client/starter/HazelcastClientStarterTest.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import java.util.concurrent.ExecutionException;
+
+import static org.junit.Assert.assertEquals;
+
+@RunWith(HazelcastSerialClassRunner.class)
+@Category(NightlyTest.class)
+public class HazelcastClientStarterTest {
```
Comment (ihsandemir) [inline hazelcast-client/src/test/java/com/hazelcast/client/starter/HazelcastClientStarterTest.java]: Are these tests directly related to this PR?

File: hazelcast/src/main/java/com/hazelcast/client/impl/ClientPartitionListenerService.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    private final NodeEngineImpl nodeEngine;
+
+    ClientPartitionListenerService(NodeEngineImpl nodeEngine) {
+        this.nodeEngine = nodeEngine;
+    }
+
+    public void onPartitionStateChange() {
+
+        for (Map.Entry<ClientEndpoint, Long> entry : partitionListeningEndpoints.entrySet()) {
+            ClientMessage clientMessage = getPartitionsMessage();
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/impl/ClientPartitionListenerService.java]: can this line be put outside the loop.

File: hazelcast/src/main/java/com/hazelcast/client/impl/ClientPartitionListenerService.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+            ClientEndpoint clientEndpoint = entry.getKey();
+            Connection connection = clientEndpoint.getConnection();
+            connection.write(clientMessage);
+        }
+    }
+
+    private ClientMessage getPartitionsMessage() {
+        Collection<Map.Entry<Address, List<Integer>>> partitions = getPartitions();
+        ClientMessage clientMessage = ClientAddPartitionListenerCodec.encodePartitionsEvent(partitions);
+        clientMessage.setCorrelationId(clientMessage.getCorrelationId());
```
Comment (ihsandemir) [inline hazelcast/src/main/java/com/hazelcast/client/impl/ClientPartitionListenerService.java]: ?

File: hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java (hunk 1 above)
Comment (sancar) [reply hazelcast-client/src/main/java/com/hazelcast/client/impl/HazelcastClientInstanceImpl.java]: In a distributed environment does it matter to get the partition table changes right before you connected ? Since things will work if you connected X ms later and get the changes right after, the delay between creating the connection and getting the changes should not matter right ?

File: hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientPartitionServiceImpl.java (hunk 2 above)
Comment (sancar) [reply hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientPartitionServiceImpl.java]: Yes, you are right. We need to wait for response to see if it is failed. 
But if I add I will broke the compatibility with some servers. Because we have released servers that does not return any response when client message is not supported. Related issue if you are wondering. 
https://github.com/hazelcast/hazelcast/issues/10196

Comment (sancar) [reply hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientPartitionServiceImpl.java]: A second way can be write the code according to server version coming with response in authentication.

File: hazelcast-client/src/test/java/com/hazelcast/client/starter/HazelcastClientStarterTest.java (hunk 5 above)
Comment (sancar) [reply hazelcast-client/src/test/java/com/hazelcast/client/starter/HazelcastClientStarterTest.java]: These tests were under hazelcast instead of hazelcast-client. Because the support of client starter with version was under hazelcast. Since client classes are not available there there was usage of reflection and class loading. And the support of client config was missing for clients. 
I cleaned the unnecessary reflection usage and added client config support. While doing that,  I  had to move client related classes and tests to hazelcast-client package.

**Comment by sancar** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/client/impl/ClientPartitionListenerService.java (hunk 6 above)
Comment (sancar) [reply hazelcast/src/main/java/com/hazelcast/client/impl/ClientPartitionListenerService.java]: Since we write to connection it does not directly write to socket but its queued, client message will be shared, and client messages needs to have different correlation id . that is why I need to create new one on each iteration.

**Comment by sancar** [comment]
run-lab-run

File: hazelcast-client/src/test/java/com/hazelcast/client/starter/HazelcastClientStarter.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+public class HazelcastClientStarter {
+
+    private HazelcastClientStarter() {
+    }
+
+    public static HazelcastInstance newHazelcastClient(String version, boolean enterprise) {
+        return newHazelcastClient(version, null, enterprise);
+    }
+
+    public static HazelcastInstance newHazelcastClient(String version, ClientConfig clientConfig, boolean enterprise) {
```
Comment (ihsandemir) [inline hazelcast-client/src/test/java/com/hazelcast/client/starter/HazelcastClientStarter.java]: enterprise flag does not look very relevant for OSS code, does it?

File: pom.xml [hunk 9]
```diff
@@ -38,7 +38,7 @@
 
     <properties>
         <main.basedir>${project.basedir}</main.basedir>
-        <client.protocol.version>1.5.0-5</client.protocol.version>
+        <client.protocol.version>1.5.0-6</client.protocol.version>
```
Comment (ihsandemir) [inline pom.xml]: 1.5.0-7 ?

File: hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientPartitionServiceImpl.java [hunk 10]
```diff
@@ -93,56 +123,44 @@ private void getPartitionsBlocking() {
                 e.printStackTrace();
             }
         }
+        while (partitionCount == 0 && client.getConnectionManager().isAlive());
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientPartitionServiceImpl.java]: This method is not trying to guarantee to take the partitions from last RefreshTask. It is trying to make sure partition table initialised. I will change the name of the method to emphasise that.

File: hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientPartitionServiceImpl.java [hunk 11]
```diff
@@ -93,56 +123,44 @@ private void getPartitionsBlocking() {
...
+            if (!partitionStateVersionExist || partitionStateVersion >= lastPartitionStateVersion) {
+                for (Map.Entry<Address, List<Integer>> entry : partitions) {
+                    Address address = entry.getKey();
+                    for (Integer partition : entry.getValue()) {
+                        this.partitions.put(partition, address);
+                    }
+                }
+                partitionCount = this.partitions.size();
+                lastPartitionStateVersion = partitionStateVersion;
+                if (logger.isWarningEnabled()) {
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientPartitionServiceImpl.java]: This should of been finest. I mistakenly forget it there. 
and the check is to make sure there are no unnecessary String creations.

File: hazelcast-client/src/test/java/com/hazelcast/client/starter/HazelcastClientStarter.java (hunk 8 above)
Comment (sancar) [reply hazelcast-client/src/test/java/com/hazelcast/client/starter/HazelcastClientStarter.java]: This is the infrastructure to test different versions of instances, and it is used by both open source and enterprise. We can abstract this part probably. This is not a new code by the way, I just moved it from `hazelcast` to `hazelcast-client` because it was not relevant there :).

File: hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientPartitionServiceImpl.java [hunk 12]
```diff
@@ -93,56 +123,44 @@ private void getPartitionsBlocking() {
...
     }
 
-    private boolean getPartitions() {
-        Connection connection = getOwnerConnection();
-        if (connection == null) {
-            return false;
+    private boolean processPartitionResponse(Collection<Map.Entry<Address, List<Integer>>> partitions,
+                                             int partitionStateVersion,
+                                             boolean partitionStateVersionExist) {
+        if (partitions.size() == 0) {
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientPartitionServiceImpl.java]: This is a leftover from a bug hunt. It is not possible to get size 0. I will remove that.

File: pom.xml (hunk 9 above)
Comment (sancar) [reply pom.xml]: For this development 1.5.0-6 is enough.

File: hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientPartitionServiceImpl.java [hunk 13]
```diff
@@ -93,56 +123,44 @@ private void getPartitionsBlocking() {
...
-            ClientGetPartitionsCodec.ResponseParameters response = ClientGetPartitionsCodec.decodeResponse(responseMessage);
-            if (response == null) {
-                return false;
-            }
-            return processPartitionResponse(response);
-        } catch (Exception e) {
-            if (client.getLifecycleService().isRunning()) {
-                logger.warning("Error while fetching cluster partition table!", e);
+        synchronized (lock) {
+            if (!partitionStateVersionExist || partitionStateVersion >= lastPartitionStateVersion) {
```
[the remark this answers is not in the document]
Comment (sancar) [reply hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientPartitionServiceImpl.java]: Changing this also.

**Comment by sancar** [comment]
run-lab-run

**Comment by sancar** [comment]
run-lab-run

**Comment by sancar** [comment]
run-lab-run

File: hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientPartitionServiceImpl.java [hunk 14]
```diff
@@ -67,30 +72,64 @@ public ClientPartitionServiceImpl(HazelcastClientInstanceImpl client) {
...
-                Thread.sleep(PARTITION_WAIT_TIME);
-            } catch (InterruptedException e) {
-                e.printStackTrace();
+                ClientMessage responseMessage = future.get();
+                ClientGetPartitionsCodec.ResponseParameters response =
+                        ClientGetPartitionsCodec.decodeResponse(responseMessage);
+                processPartitionResponse(response.partitions,
+                        response.partitionStateVersion, response.partitionStateVersionExist);
+            } catch (Exception e) {
+                if (client.getLifecycleService().isRunning()) {
```
Comment (ihsandemir) [inline hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientPartitionServiceImpl.java]: without no sleep and with current codec this causes busy loop. we need to look if we can change the codec to be retryable message. We hope that the exception will mostly be IOException or the InstanceNotActiveException, and hence hopefully will go to retry mechanism which does the sleep to slow down the busy loop.
